runtime/runtime/src/ext.rs

```diff
@@ -370,12 +370,24 @@ pub(crate) struct RuntimeContractExt<'a> {
 
 impl<'a> Contract for RuntimeContractExt<'a> {
     fn hash(&self) -> CryptoHash {
+        // For eth implicit accounts return the wallet contract code hash.
+        // The account.code_hash() contains hash of the magic bytes, not the contract hash.
+        if checked_feature!("stable", EthImplicitAccounts, self.current_protocol_version)
+            && self.account_id.get_account_type() == AccountType::EthImplicitAccount
+        {
+            // There are old eth implicit accounts without magic bytes in the code hash.
+            // Result can be None and it's a valid option. See https://github.com/near/nearcore/pull/11606
+            if let Some(wallet_contract) = wallet_contract(self.account.code_hash()) {
+                return *wallet_contract.hash();
```

Comment:
In the long run we will likely want to avoid recomputing this hash every time this function is invoked, especially in case these ethereum wallets get used heavily (as this change makes the `hash` function `O(n)` with respect to the contract size whereas it was `O(1)` before.)

Reply:
One way to implement this would be to figure out how to make `Account` store the proper code hash for the contract even if we do not store the code itself in `set_contract`.

But regardless I don't think this concern is blocking.

Reply:
My understanding is that it's still O(1). `read_contract` lazily initializes a `ContractCode` value which computes the hash during initialization. After that all subsequent calls to `hash()` use the precomputed value stored in `ContractCode`.